openvdb/openvdb/tools/Merge.h

```diff
@@ -0,0 +1,931 @@
+    bool operator()(RootT& root, size_t idx) const;
+
+    // Processes the internal nodes. Required by the NodeManager
+    template<typename NodeT>
+    bool operator()(NodeT& node, size_t idx) const;
+
+    // Processes the leaf nodes. Required by the NodeManager
+    bool operator()(LeafT& leaf, size_t idx) const;
+
+private:
+    // on processing the root node, the background value is stored, retrieve it
+    // and check that the root node has already been processed
+    const ValueT& background() const;
+
+    mutable std::vector<TreeToMerge<TreeT>> mTreesToMerge;
+    mutable const ValueT* mBackground = nullptr;
+}; // struct CsgUnionOrIntersectionOp
+
+
+template <typename TreeT>
+using CsgUnionOp = CsgUnionOrIntersectionOp<TreeT, /*Union=*/true>;
+
+template <typename TreeT>
+using CsgIntersectionOp = CsgUnionOrIntersectionOp<TreeT, /*Union=*/false>;
+
+
+/// @brief DynamicNodeManager operator to merge two trees using a CSG difference.
+/// @note This class modifies the topology of the tree so is designed to be used
+/// from DynamicNodeManager::foreachTopDown().
+template<typename TreeT>
+struct CsgDifferenceOp
+{
+    using ValueT = typename TreeT::ValueType;
+    using RootT = typename TreeT::RootNodeType;
+    using LeafT = typename TreeT::LeafNodeType;
+
+    /// @brief Convenience constructor to CSG difference a single non-const
+    /// tree from another. This constructor takes a Steal or DeepCopy tag
+    /// dispatch class.
+    template <typename TagT>
+    CsgDifferenceOp(TreeT& tree, TagT tag) : mTree(&tree, tag) { }
+    /// @brief Convenience constructor to CSG difference a single const
+    /// tree from another. This constructor requires an explicit DeepCopy tag
+    /// dispatch class.
+    CsgDifferenceOp(const TreeT& tree, DeepCopy tag) : mTree(&tree, tag) { }
+
+    /// @brief Constructor to CSG difference the tree in a TreeToMerge object
+    /// from another.
+    explicit CsgDifferenceOp(TreeToMerge<TreeT>& tree) : mTree(tree) { }
+
+    /// @brief Return the number of trees being merged (only ever 1)
+    size_t size() const { return 1; }
+
+    // Processes the root node. Required by the NodeManager
+    bool operator()(RootT& root, size_t idx) const;
+
+    // Processes the internal nodes. Required by the NodeManager
+    template<typename NodeT>
+    bool operator()(NodeT& node, size_t idx) const;
+
+    // Processes the leaf nodes. Required by the NodeManager
+    bool operator()(LeafT& leaf, size_t idx) const;
+
+private:
+    // on processing the root node, the background values are stored, retrieve them
+    // and check that the root nodes have already been processed
+    const ValueT& background() const;
+    const ValueT& otherBackground() const;
+
+    // note that this vector is copied in NodeTransformer every time a foreach call is made,
+    // however in typical use cases this cost will be dwarfed by the actual merge algorithm
+    mutable TreeToMerge<TreeT> mTree;
```

Comment:
No vector here - will this only ever work on 2 trees?

Reply:
I think this is why it isn't included in the Union/Intersect operator.  I kind of expected difference to be there, and had a comment to that effect before I noticed an intersection class.  I believe the logic is that difference isn't as well defined on a list as the first member is treated differently.

Given (T0, T1, T2, ...)
it becomes
T0 & ~( T1 | T2 | ... )
ie, union of the tail, then intersection of the inversion of the tail with the head.

Reply:
Yes, as @jmlait says, it's not totally clear what difference should do if there are multiple grids. You can achieve `T0 & ~( T1 | T2 | ...)` by doing `Difference(T0, Merge(T1, T2, ...))` so I felt it was clearer to enforce that especially as there's no performance penalty in doing so. There's currently no difference operation in my implementation of the VDB Merge SOP for this exact reason. That suggests that we would always want to keep the VDB Combine SOP (or equivalent) around in some form or another for binary operations.